### PR TITLE
ibmcloud: Flexible instance profiles

### LINKED ISSUE #1115: Add flexible instance types support for ibmcloud

### LINKED ISSUE #1153: instanceTypes `Set` with a empty string results in a non-empty list
https://github.com/confidential-containers/cloud-api-adaptor/blob/c904e24a6f108d2b93b545447c88fa7d8ea08bda/pkg/adaptor/cloud/aws/types.go#L30C1-L33C2

```go
func TestEmptyList(t *testing.T) {
	var list instanceTypes
	list.Set("")
	if len(list) != 0 {
		t.Errorf("Expect 0 length, got %d", len(list))
	}
}
```

### PR DESCRIPTION
**Pull request by jtumber-ibm**
Adds support for selecting instance based on Pod Annotations, example.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: annotations-demo
  annotations:
   io.katacontainers.config.hypervisor.machine_type: "bx2-1x4"
   io.katacontainers.config.hypervisor.default_vcpus: "3"
   io.katacontainers.config.hypervisor.default_memory: "3072"
spec:
  runtimeClassName: kata-remote
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
```

Specify the set of allowed instance profiles by setting `IBMCLOUD_PODVM_INSTANCE_PROFILE_LIST` e.g.

`IBMCLOUD_PODVM_INSTANCE_PROFILE_LIST="bz2-1x4,bz2-2x8"`

To test need to use the `quay.io/confidential-containers/runtime-payload-ci:kata-containers-latest` image 

Fixes #1115, Fixes #1153

### CODE REVIEW

File: pkg/adaptor/cloud/ibmcloud/provider.go [hunk 1]
```diff
@@ -279,6 +289,61 @@ func (p *ibmcloudVPCProvider) CreateInstance(ctx context.Context, podName, sandb
...
+	// Iterate over the instance types and populate the instanceProfilespecList
+	for _, profileType := range instanceProfiles {
+		vcpus, memory, err := p.getProfileNameInformation(profileType)
+		if err != nil {
+			return err
+		}
+		instanceProfilespecList = append(instanceProfilespecList, cloud.InstanceTypeSpec{InstanceType: profileType, VCPUs: vcpus, Memory: memory})
+	}
+
+	// Sort the instanceProfilespecList by Memory and update the serviceConfig
```
Comment (stevenhorsman) [inline pkg/adaptor/cloud/ibmcloud/provider.go:321]: Minor nit: instanceProfilespecList -> instanceProfile*S*pecList

**Review by stevenhorsman — APPROVED** [review]
The code looks good to me, just retesting the behaviour now (which worked for me on Friday)

**Comment by stevenhorsman** [comment]
I've tested this locally with the following set-up:
JT's build of the CAA:
```
images:
- name: cloud-api-adaptor
  newName: jamesnelson/cloud-api-adaptor
  newTag: e18e96ebc1195aa5144b0fc33d2a54acd253ca57
```
and the following instance set up:
```
cat install/overlays/ibmcloud/kustomization.yaml | grep -i instance
  - IBMCLOUD_PODVM_INSTANCE_PROFILE_NAME="bx2-2x8"
  - IBMCLOUD_PODVM_INSTANCE_PROFILE_LIST="bx2-2x8,cx2-2x4,mx2-2x16"
```
I then created pod for a variety of test cases:

**Default**
My pod spec had no annotations:
```
apiVersion: v1
kind: Pod
metadata:
  name: no-annotations
spec:
  runtimeClassName: kata-remote
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
```
This is expected to use the default instance type `bx2-2x8`
  
**Instance type**
I specified a valid instance type
```
apiVersion: v1
kind: Pod
metadata:
  name: annotations-instance-type
  annotations:
   io.katacontainers.config.hypervisor.machine_type: "cx2-2x4"
spec:
  runtimeClassName: kata-remote
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
```
This is expected to use the annotation instance type: `cx2-2x4`

**Invalid instance type**
I specified a valid instance type
```
apiVersion: v1
kind: Pod
metadata:
  name: annotations-invalid-instance-type
  annotations:
   io.katacontainers.config.hypervisor.machine_type: "bx2-2x4"
spec:
  runtimeClassName: kata-remote
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
```
This should fail with a good error

**CPU and memory picked**
This pod spec had cpu and memory selected that were within the bounds of the options:
```
apiVersion: v1
kind: Pod
metadata:
  name: annotations-cpu-mem
  annotations:
   io.katacontainers.config.hypervisor.default_vcpus: "2"
   io.katacontainers.config.hypervisor.default_memory: "12288"
spec:
  runtimeClassName: kata-remote
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
```
This is expected to create an instance of `mx2-2x16` as the smallest instance profile that meets the requriement

**Memory too large**
I specified a memory larger than any of the profiles
```
apiVersion: v1
kind: Pod
metadata:
  name: annotations-too-big-mem
  annotations:
   io.katacontainers.config.hypervisor.default_vcpus: "2"
   io.katacontainers.config.hypervisor.default_memory: "18432"
spec:
  runtimeClassName: kata-remote
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
```
This should fail with a good error

**CPU too large**
I specified vCPUs larger than any of the profiles:
```
apiVersion: v1
  runtimeClassName: kata-remote
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
apiVersion: v1
kind: Pod
metadata:
  name: annotations-too-big-cpu
  annotations:
   io.katacontainers.config.hypervisor.default_vcpus: "3"
   io.katacontainers.config.hypervisor.default_memory: "12288"
spec:
  runtimeClassName: kata-remote
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
```
This should fail with a good error

I ran all these pods and saw the following pod status:
```
kubectl get pods
NAME                                READY   STATUS              RESTARTS   AGE
annotations-cpu-mem                 1/1     Running             0          32m
annotations-instance-type           1/1     Running             0          32m
annotations-invalid-instance-type   0/1     ContainerCreating   0          4m5s
annotations-too-big-cpu             0/1     ContainerCreating   0          32m
annotations-too-big-mem             0/1     ContainerCreating   0          32m
no-annotations                      1/1     Running             0          32m
```
which is a good start

Looking at the instances created:
```
ID                                          Name                                       Status    Reserved IP   Floating IP     Profile    Image                                VPC                            Zone       Resource group
02r7_246cf1a3-8391-40bb-918c-39c66d51b4e6   podvm-annotations-cpu-mem-441368b0         running   10.249.64.7   -               mx2-2x16   podvm-b5b55f2-dirty-amd64            sh-sm-cluster-2023-07-06-vpc   ca-tor-2   default
02r7_5db3edb7-f920-4ab6-b573-50e559764b56   podvm-annotations-instance-type-545f8ea2   running   10.249.64.9   -               cx2-2x4    podvm-b5b55f2-dirty-amd64            sh-sm-cluster-2023-07-06-vpc   ca-tor-2   default
02r7_54f513d4-0b4e-4e3e-bdf4-2373fe65f79b   podvm-no-annotations-446dace6              running   10.249.64.8   -               bx2-2x8    podvm-b5b55f2-dirty-amd64            sh-sm-cluster-2023-07-06-vpc   ca-tor-2   default
```
it looks like all the correct instances were selected.
Looking at the failures and describing the pods shows:
```
Name:                annotations-too-big-mem
Events:
  Type     Reason                  Age                   From     Message
  ----     ------                  ----                  ----     -------
  Warning  FailedCreatePodSandBox  115s (x318 over 71m)  kubelet  Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to get instance type based on vCPU and memory annotations: no instance type found for the given vcpus (2) and memory (18432): unknown
```

```
Name:                annotations-too-big-cpu
Events:
  Type     Reason                  Age                   From     Message
  ----     ------                  ----                  ----     -------
  Warning  FailedCreatePodSandBox  106s (x318 over 71m)  kubelet  Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: Number of cpus 3 specified in annotation default_vcpus is greater than the number of CPUs 2 on the system: unknown
```

```
Name:                annotations-invalid-instance-type
Events:
  Type     Reason                  Age                    From               Message
  ----     ------                  ----                   ----               -------
  Normal   Scheduled               43m                    default-scheduler  Successfully assigned default/annotations-invalid-instance-type to sh-sm-cluster-2023-07-06-node-1
  Warning  FailedCreatePodSandBox  3m19s (x182 over 43m)  kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to verify instance type: requested instance type ("bx2-2x4") is not part of supported instance types list: unknown
```
Which all seem to give a helpful message. Base on this testing I'm very happy with this PR. Just one thing can you create an issue to automate similar test cases that I have down to make sure we don't forget? Thanks!

**Comment by bpradipt** [comment]
@jtumber-ibm can you please update azure provider with the fix for #1153 ?

**Pull request merged**